README.md
*last updated on October 27/2015*

*Can enable the panel Overlay with panelOverlay: true*

*When you need the off Flag you can call it with offFlag: true* 



# Aviation Project

This branch is for combining front-end libraries and features used in the edX/Aviation project at CDOT

## References to works used in this project and
### JavaScript and jQuery libraries required by this library:

    SlickQuiz: https://github.com/pkuzhel/SlickQuiz
        Original library: https://github.com/jewlofthelotus/SlickQuiz
    TimerJS: https://github.com/pkuzhel/timer.js
        Original library: https://github.com/husa/timer.js
    PopcornJS: http://cdn.popcornjs.org/code/dist/popcorn.min.js || https://github.com/mozilla/popcorn-js
    jQuery Flight Indicators: https://github.com/uw-ray/Skyhawk-Flight-Instruments
        Continued from: https://github.com/echanna/jQuery-Flight-Indicators/tree/cdot_with_turn_coordinator
            Original: https://github.com/sebmatton/jQuery-Flight-Indicators
    JS CSV Parser (PapaParse): http://papaparse.com/
    Resize Detection (CSS Element Queries): https://github.com/marcj/css-element-queries
	jQuery Pulse Animation Plugin: https://github.com/jsoverson/jquery.pulse.js


## Slide Creation Library for the internal use of the edX team at CDOT

In order to have this version of the createSlide library working you will need the following files: (these files should be already organized inside Project/aviation/)

    CDOT-EDX/Project/library/static/js/content/createSlide.js
    CDOT-EDX/Project/library/statis/js/*
    CDOT-EDX/Project/library/statis/css/*

createSlide now supports one options object being passed versus multiples...

therefore use: `var anySlide = new AVIATION.Slide( options )` (if ComplexHTML xblock isn't being used)

*NOTE: The example below doesn't use a properly formatted JSON, make sure to use proper JSON*

*This can be validated easily using a tool such as: http://jsonlint.com/*

Acceptable options and defaults are as follows:

```javascript
    serverBaseUrl:      window.location.protocol + "//" + window.location.host + "/",
    apacheServerBaseUrl:window.location.protocol + "//" + window.location.host + ":25080/",
    setPanel:           false, // makes the first line of the first CSV set panel before CSV plays
    parentSlide:        {},
    development         false, // if running inside a devstack set to true
    showAvatars:        false,
    showSlideControls:  true,
    showStatus:         true,
    showRemediationOnSuccess: true, //show all remediations on Correct answer
    showRemediationOnFail: true, //show all remediations on inCorrect answer
    showRemediationOne: true, //will show remediation for selected answer (showRemediationOnSuccess or showRemediationOnFail must be set to true) !!can only be used only with showRemediationOnSuccess or showRemediationOnFail
    showControls:       true,
    showBorder:         true, // this applies to the border around the content
    autoplay:           true, // proceeds to the next AUDIO automatically after current is finished
    autoRedirect:       false, // proceeds to the next SLIDE automatically after current is finished
    noAudio:            false, // in case the slide doesn't have any audio content to control the content creation
    noCSV:              false, // indicates that the instrument panel should build even if a csv file isn't specified in media object
    enableModals:       false,
    enableHighlights:   false,
    hiddenHighlights:   false,
    enableButtons:      true,
    enableSlider:       false,
    enablePanel:        false, // enable full panel unless "panelType" option is set
    panelType:          { "attitude": "attitude" }, // object that lists 1 or more instruments to display (if not set, full panel)
    panelOverlay:       false,
    offFlag:            false,
    readOnlySlider:     true,
    container:          "#slideContainer", // where should we look for the slideContainer?
    statusId:           "#statusBar", // whats the id that we should create for statusBar?
    headerId:           "#header",
    footerId:           "#footer",
    bodyId:             "#body",
    headerId:           "#slideHeader",
    footerId:           "#slideFooter",
    quizId:             "#slideQuiz",   // id of div that contains all quizzes
    quizContainerClass: "cdot_quiz_container", // class thats appended to each individual quiz
    

    // new optinal+experimental additions below
    panelId: "#flightInstruments",
    panelHighlightsId: "#panelHighlightContainer",
    instStatusId1: "#instStatus1",
    instStatusId2: "#instStatus2",
    scanningPatternArray: [ 1, 0, 1, 2, 1, 5, 1, 4, 1, 3 ],
    minScan: 1,
    pulseCorrectProp: {
      borderWidth: '5px',
      borderColor: 'green',
      borderRadius: '5px'
    },
    pulseWrongProp: {
      borderWidth: '5px',
      borderColor: 'red',
      borderRadius: '5px'
    },
    pulseInstrumentSettings: {
      duration: 500,
      pulses: 1
    },
    pulseProperties: {
      backgroundColor: "#D0FAEE",
      borderRadius: "10px"
    },
    pulseSettings: {
      duration: 3000,
      pulses: 1
    },
    
    // both options below are used to pull audio/images/csvs from the correct server without running into cross-origin restrictions
    serverBaseUrl: window.location.protocol + "//" + window.location.host + "/", //can be overwritten by a string of the host url
    apacheServerBaseUrl: window.location.protocol + "//" + window.location.host + ":25080/",

    **possibly deprecated and moved inside slideContent:
    **  advanceWith:        "audio", // other options -> "highlight", "timer", "button", "instruments"

    continueId:         "", // string specifying the next id to redirect to
    backId:             "", // string specifying the previous id to redirect to
    
    panelOptions: { // options passed on to flightInstruments library
        size : 200,             // Sets the size in pixels of the indicator (square)
        showBox : true,         // Sets the visibility of the box behind the instruments
        showScrews: true,       // Sets the visibility of the four screws around the instruments
        airspeed: 0,            // Air speed in knots for an air speed indicator
        roll : 0,               // Roll angle in degrees for an attitude indicator
        pitch : 0,              // Pitch angle in degrees for an attitude indicator
        off_flag: false         // Off flag for an attitude indicator
        altitude: 0,            // Altitude in feets for an altimeter indicator
        pressure: 30,           // Pressure in inHg for an altimeter indicator
        turn: 0,                // Turn direction for turn coordinator
        slip: 0,                // Slip ball position for turn coordinator (0 to 1; 0.5 is middle)
        heading: 0,             // Heading angle in degrees for an heading indicator
        beaconone: 0,           // Angle of first beacon on the heading indicator
        beacononeshow: true,    // Sets the visibility of the first beacon on the heading indicator
        beacontwo: 0,           // Angle of second beacon on heading indicator
        beacontwoshow: true,    // Sets the visibility of the second beacon on the heading indicator
        vario: 0,               // Variometer in 1000 feets/min for the variometer indicator
        img_directory : 'img/'  // The directory where the images are saved to
    },

    quizzes: [ // options passed on to the slickQuiz library
    // most explanations to these can be found at: https://github.com/jewlofthelotus/SlickQuiz
      {
        json: {
          "info": {
            "name":    "Answer the following questions to the best of your ability.",
            "main":    "",
            "results": "",
            "level1":  "",
            "level2":  "",
            "level3":  "",
            "level4":  "",
            "level5":  "" // no comma here
          },
          "questions": [
            { // Question 1 - Multiple Choice, Single True Answer
              "q": "Now speak the correct clearance: >>AS12, I read back",
              "a": [
                {"option": "Climb on runway heading to 3000 ft, turning right onto 125 degrees", "correct": "", "reason": ""},
                {"option": "Climb on 135 degrees to 1500 ft, then climb to 3000 ft", "correct": "",  "reason": ""},
                {"option": "Climb on runway heading to 1500 ft, turning right onto 125 degrees", "correct": "",  "reason": ""},
                {"option": "Climb on runway heading to 1500 ft, turning right onto 135 degrees. Continue climb to 3000 feet to complete the correct read back.",  "correct": "",  "reason": ""} // no comma here
              ],
              "correct": "<p><span>Control, correct, AS12 is clear to take-off</span></p>",
              "incorrect": "<p><span>That's not correct.  Ask for the clearance to be repeated.</span></p>" // no comma here
            }
          ],
        },
        // not mentioned on SlickQuiz page but added by edX CDOT team
        "showRemediationOnSuccess": true,
        "showRemediationOnFail": true,
        "animationCallbacks": {
          "checkAnswer": (function(){
            console.log("do something");
            $(anySlide).trigger("completedQuiz");
          })
        }
      }
    ],
    avatars: 
    {
      tom: {
        ~~open: "//online.cdot.senecacollege.ca:25080/aviation/img/tomOpen.png",~~
        ~~close: "//online.cdot.senecacollege.ca:25080/aviation/img/tomClose.png"~~
        // since server url options have been added, we can give relative locations to resources
        open: "aviation/img/tomOpen.png",
        close: "aviation/img/tomClose.png"
      },
      jane: {
        open: "aviation/img/janeOpen.png",
        close: "aviation/img/janeClose.png"
      }
    },
    
    highlights:
    [
      {
        id: "asi",
        orderNumber: 0,
        name: "Airspeed Indicator (ASI)",
        height: "50%",
        classes: ["col-xs-4"],
        border : "7px ridge yellow",
        parent: 'panel'
      },
      {
        id: "ai",
        orderNumber: 1,
        name: "Attitude Indicator (AI)",
        height: "50%",
        classes: ["col-xs-4"],
        border : "7px ridge yellow",
        parent: 'panel'
      }, 
      {
        id: "alt",
        orderNumber: 2,
        name: "Altimeter (ALT)",
        height: "50%",
        classes: ["col-xs-4"],
        border : "7px ridge yellow",
        parent: 'panel'
      },
      {
        id: "tc",
        orderNumber: 3,
        name: "Turn Coordinator (TC)",
        height: "50%",
        classes: ["col-xs-4"],
        border : "7px ridge yellow",
        parent: 'panel'
      },
      {
        id: "hi",
        orderNumber: 4,
        name: "Heading Indicator (HI)",
        height: "50%",
        classes: ["col-xs-4"],
        border : "7px ridge yellow",
        parent: 'panel'
      },
      {
        id: "vsi",
        orderNumber: 5,
        name: "Vertical Speed Indicator (VSI)",
        height: "50%",
        classes: ["col-xs-4"],
        border : "7px ridge yellow",
        parent: 'panel'
      },
    ],
    },
    // This is for buttons that we want to use on the slide
    buttons: [
        {
          "id": "btn0",
          "title": 'I am a button',
          "classes": ["btn", "btn-default"],
          "orderNumber": 0,
          "action": "(function(){ console.log('button of someId is executing'); })"
        },
        {
          "id": "btn1",
          title: 'another button',
          "orderNumber": 1
        },
        {
          "id": "topple",
          "title": "TOPPLE",
          "mediaIndex": 3 // mediaIndex will invoke "play" on mediaFile at index 3 once button is clicked
        }
    ],
    slideContent: [{
        title: {html: "No content provided", classes: ["col-md-12"] }, 
        content: {html: "<p<Check your <b>slideContent</b> object</p>", classes: ["text-center"] },
        image: {src: "/someSource/image.png", classes: ["imageClass"] },
        highlights: [ 
                        0 , 
                        { index: 2, onclick: (function(){ console.log("highlight onclick");}) }, 
                        5 
        ],
        // this is to hide / show buttons just like we do with highlights
        buttons: [
                    0 , 
                    { "index": 2, "disable": true, "onclick": "(function(){ console.log('button onclick');})" }, 
                    3 
        ],
        // set slider to any throttle possition from 0-4
        slider: 2,
        "media":{
            "type": "audio", // could be csv/button/highlight/quiz
            "index": 0,
            "second" 10
        },
        callback: (function(){ console.log("this is a callback function"); }),
        "advanceWith": {
            "action": "pattern",
            // for both properties below, anything to do with the contentIndex will only rebuild the content for that
            // index and not replay the media associated with the content
            // BUT, anything to do with mediaIndex, will restart the media and run all of the content associated with that media again
            // "<-"/previous btn and "->"/forward btn on the slides skip mediaIndices NOT contentIndices
            "onSuccess": { ** optional if needed and currently under development
                "status": "Good job, let's move on...",
                // if onSuccess content is not the next slideContent....
                "contentIndex": 3 // optional, which index should we jump to in slideContent OR use mediaIndex instead
                "mediaIndex": 2 // mediaIndex will jump to the beginning of a media at set index and run all the content from that
                "callback": "(function(){ console.log('You did smth right'); })" // optional
            },
            "onFail": { ** optional if needed and currently under development
                "index": 2, // which altMediaFile should be played (currently sound only)
                "status": "That's not correct. Let's try this again", // what msg should we display in the status bar?
                "resetContentIndex": 0 // which slideContent index do we need show as altMedia starts to play?
                "resetMediaIndex": 0 // which media index do we need to play after altMedia?
            },
            "content": true // if there is content to that should be triggered during the scanning pattern
        }
    },
    {
      	"media":{
            "type": "timer", 
            "index": 1,
            "second" 1 // timer is a count down , starts at 10 and ends at 1 
    	}, 
    	"buttons": [0],
        "content": {
	   "html": "UH OH, Press that button yo!"
        },
        "advanceWith": {"type": "button", "index": 0}
    }, 
    {
        //... another slide content can go here to be triggered at a different time of the audio
        title: {html: "This is a different title / slide"},
        slider: 3,
        "media": {
            "type": "csv",
            "line": 100,
            "index": 2
        }
        "advanceWith": {
            "type": "button", // highlight, audio, csv
            "index": 0,
            "onFail": {
                "index": 0,
                "resetIndex": 1
            }
        }
    },
    // PAUSE / PLAY example for CSV
    {
        "media":{
            "index": 2,
            "type": "csv",
            "line": 1000,
        },
        "action": {
          "type": "pause",
          "index": 2
        },
        "advanceWith": {
            "type": "button",
            "index": 1
        },
    },
    {
        "media": {
          "type": "button",
          "index": 1
        },
        "action":{
          "type": "play",
          "index": 2 // can only be a mediaIndex (not contentIndex)
          "line": 1001 // in this case its a csv
          // in the case of an audio we can specify "second": 10.1 and not 'line'
        }
    },
    // PAUSE / PLAY example for AUDIO
    {
        "media":{
            "index": 2,
            "type": "audio",
            "second": 7,
        },
        "action": {
          "type": "pause",
          "index": 2
        },
        "advanceWith": {
            "type": "button",
            "index": 1
        },
    },
    {
        "media": {
          "type": "button",
          "index": 1
        },
        "action":{
          "type": "play",
          "index": 2 // can only be a mediaIndex (not contentIndex)
          "second": 7.1 // in this case its an audio
        }
    }],
    // default models that we want to be set-up in the background
    modals: [{ 
              id: "ai", 
              title: "Attitude Indicator", 
              content: {html: "This is the MODAL custom HTML" },
            }, 
            {
              id: "alt", 
              title: "Altimeter", 
            }
    ],
    // these audioFiles are for the slide only
    mediaFiles: [ 
        // since server url options have been added, we can give relative locations to resources omitting the https://online...../
        {type: "audio", src: "someRelativeURL/toAudioFile"},
        {type: "audio", src: "anotherURL/toOtherFile"},
        {type: "csv", src: "someURL/flightData"},
        {type: "timer", "duration": 10}
    ],
    altMediaFiles: [ // media files that will be used in alternative content (in case a student answered a question wrong)
        {type: "audio", src: "someURL/toAudioFile"},
        {type: "audio", src: "anotherURL/toOtherFile"},
    ]
```

The rest of this README is being edited at this moment
**More options and capabilities are being added all the time**
**Everything below is just notes of the developer for the developer**

## Syntax and actions available for content

    content - can include any custom html if needed
        html (string/html)
        classes (string array)
        action (string) - available actions: remove, replace, append **replace is default**


## Global Events available for triggering

    play(playWhat, whichIndex)
        - with no parameters passed, will start playing slide audios (and stop all other audios)
        - with "modal" passed, will start playing the modal audio
        - with "extra" passed, will start playing the extra audio within slides

    next

    previous

    pause

    stop

    replay


    Slide class methods:
        
        new Slide(options, slideContent, audioFiles)
        create( options )
        destroy()
        constructor()
        _initSimple()
        buildHeader( parent, content, callback, clearTitle)
        buildSlide()
        buildAvatars(parent, avatar, callback)
        buildContent( correctAudio, index, outerIndex, clearContent )
        buildSlideControls()
        insertLineBreak(parent)
        buildSlideAudios(modalAudios, parent, modalIndex)
        buildStatusBar()
        buildFooter()
        buildCourseControls()
        buildHighlights(index, modalHighlight)
        buildModals()
        activateTimer()
        resetTimer()
        playCurrent()
        pauseCurrent()
        playPrevious()
        playNext()
        replayAll()
        replayCurrent()
        buttonOnClickEvents()
        initButtonEvents()
        initAudioEvents()
        checkSlideControlButtons()
        activateTimer( seconds, isAuto )
        initCourseButtonEvents()
        initAudioEvents()
        checkSlideControlPlayButtons( action );
        setStatus ( action ) - set status bar depending on action: play, pause, or display the actual "action"
        activateSlide()
        resetSlide()
        redirectToPage( pageId ) - use instead of "jump_to_id"

    Slide class properties:

        type
            (string) - simple/highlights)
        options
            development (if running inside a devstack, set to true)
            autoplay (boolean)
            showAvatars (boolean)
            showSlideControls (boolean)
            showStatus (boolean)
            showControls (boolean)
            showBorder (boolean)
            continueId (string)
            backId (string)
            noAudio (boolean)
            enableHighlights (boolean)
            hiddenHighlights (boolean)
            enableModals (boolean)
        activeIndex
            (integer which keeps track of the active audio of the slide)
        slideContent (object array)
            avatar (object array)
                type (string) - available avatars: [open, close, ...]
                position - left / right / hide
                character - Tom / Jane
            title - can include any custom html if needed
                html (string/html)
                classes (string array)
                action (string) - available actions: remove 

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased][unreleased]
### Changed

## [0.1.1] - 2015-09-23
### Changed
- improvements to advanceWith, fixed bug advancing after a CSV end
- CSV end event has been simplified, advanceWith check moved into checkActionable and not inside papaComplete
- got rid of VSI transform 1.0s (to give faster feedback according to Gordon)
- fix to huge bug with multiple csvs and sequential playing
- interval declaration inside papaComplete is now moved inside inner anonymous function
- csv indices are attacahed to papaconfig so that they are pushed in the appropriate order instead of the first csv being parsed

### Added
- advanceWith now can have an array of types and indices
- papaCueEnd + papaCuePause to be able to execute custom functions on those actions of the CSV

## [0.0.1] - 2015-09-19
### Added
- This CHANGELOG file to keep track of versioning and changes to the library
- Ability to disable/enable btns/hlights (haven't tested how this affects quizzes)
- events: correctAdvance, wrongAdvance, playAltIndex, contentNext, slideEnd, altEnd, nextMedia

### Changed
- changed bootstrap approach to instruments
- activeIndex to be mediaActiveIndex and contentActiveIndex respectively (seems succcessful)
- next event

[0.1.1]: https://github.com/CDOT-EDX/Project/commit/2f13cf3722783c07914e454d8fbb369279f019ad
[0.0.1]: https://github.com/CDOT-EDX/Project/commit/9a3841a874c4a8824a0eaf3c905af172c753d3ef

svg/jQueryRotate.js
// VERSION: 2.3 LAST UPDATE: 11.07.2013
/* 
 * Licensed under the MIT license: http://www.opensource.org/licenses/mit-license.php
 * 
 * Made by Wilq32, wilq32@gmail.com, Wroclaw, Poland, 01.2009
 * Website: http://code.google.com/p/jqueryrotate/ 
 */
 
(function($) {
    var supportedCSS,supportedCSSOrigin, styles=document.getElementsByTagName("head")[0].style,toCheck="transformProperty WebkitTransform OTransform msTransform MozTransform".split(" ");
    for (var a = 0; a < toCheck.length; a++) if (styles[toCheck[a]] !== undefined) { supportedCSS = toCheck[a]; }
    if (supportedCSS) {
      supportedCSSOrigin = supportedCSS.replace(/[tT]ransform/,"TransformOrigin");
      if (supportedCSSOrigin[0] == "T") supportedCSSOrigin[0] = "t";
    }
 
    // Bad eval to preven google closure to remove it from code o_O
    eval('IE = "v"=="\v"');
 
    jQuery.fn.extend({
        rotate:function(parameters)
        {
          if (this.length===0||typeof parameters=="undefined") return;
          if (typeof parameters=="number") parameters={angle:parameters};
          var returned=[];
          for (var i=0,i0=this.length;i<i0;i++)
          {
            var element=this.get(i);	
            if (!element.Wilq32 || !element.Wilq32.PhotoEffect) {
 
              var paramClone = $.extend(true, {}, parameters); 
              var newRotObject = new Wilq32.PhotoEffect(element,paramClone)._rootObj;
 
              returned.push($(newRotObject));
            }
            else {
              element.Wilq32.PhotoEffect._handleRotation(parameters);
            }
          }
          return returned;
        },
        getRotateAngle: function(){
          var ret = [];
          for (var i=0,i0=this.length;i<i0;i++)
          {
            var element=this.get(i);	
            if (element.Wilq32 && element.Wilq32.PhotoEffect) {
              ret[i] = element.Wilq32.PhotoEffect._angle;
            }
          }
          return ret;
        },
        stopRotate: function(){
          for (var i=0,i0=this.length;i<i0;i++)
          {
            var element=this.get(i);	
            if (element.Wilq32 && element.Wilq32.PhotoEffect) {
              clearTimeout(element.Wilq32.PhotoEffect._timer);
            }
          }
        }
    });
 
    // Library agnostic interface
 
    Wilq32=window.Wilq32||{};
    Wilq32.PhotoEffect=(function(){
 
      if (supportedCSS) {
        return function(img,parameters){
          img.Wilq32 = {
            PhotoEffect: this
          };
 
          this._img = this._rootObj = this._eventObj = img;
          this._handleRotation(parameters);
        }
      } else {
        return function(img,parameters) {
          this._img = img;
          this._onLoadDelegate = [parameters];
 
          this._rootObj=document.createElement('span');
          this._rootObj.style.display="inline-block";
          this._rootObj.Wilq32 = 
            {
              PhotoEffect: this
            };
          img.parentNode.insertBefore(this._rootObj,img);
 
          if (img.complete) {
            this._Loader();
          } else {
            var self=this;
            // TODO: Remove jQuery dependency
            jQuery(this._img).bind("load", function(){ self._Loader(); });
          }
        }
      }
    })();
 
    Wilq32.PhotoEffect.prototype = {
      _setupParameters : function (parameters){
        this._parameters = this._parameters || {};
        if (typeof this._angle !== "number") { this._angle = 0 ; }
        if (typeof parameters.angle==="number") { this._angle = parameters.angle; }
        this._parameters.animateTo = (typeof parameters.animateTo === "number") ? (parameters.animateTo) : (this._angle); 
 
        this._parameters.step = parameters.step || this._parameters.step || null;
        this._parameters.easing = parameters.easing || this._parameters.easing || this._defaultEasing;
        this._parameters.duration = parameters.duration || this._parameters.duration || 1000;
        this._parameters.callback = parameters.callback || this._parameters.callback || this._emptyFunction;
        this._parameters.center = parameters.center || this._parameters.center || ["50%","50%"];
        if (typeof this._parameters.center[0] == "string") {
          this._rotationCenterX = (parseInt(this._parameters.center[0],10) / 100) * this._imgWidth * this._aspectW;
        } else {
          this._rotationCenterX = this._parameters.center[0];
        }
        if (typeof this._parameters.center[1] == "string") {
          this._rotationCenterY = (parseInt(this._parameters.center[1],10) / 100) * this._imgHeight * this._aspectH;
        } else {
          this._rotationCenterY = this._parameters.center[1];
        }
 
        if (parameters.bind && parameters.bind != this._parameters.bind) { this._BindEvents(parameters.bind); }
      },
      _emptyFunction: function(){},
      _defaultEasing: function (x, t, b, c, d) { return -c * ((t=t/d-1)*t*t*t - 1) + b }, 
      _handleRotation : function(parameters, dontcheck){
        if (!supportedCSS && !this._img.complete && !dontcheck) {
          this._onLoadDelegate.push(parameters);
          return;
        }
        this._setupParameters(parameters);
        if (this._angle==this._parameters.animateTo) {
          this._rotate(this._angle);
        }
        else { 
          this._animateStart();          
        }
      },
 
      _BindEvents:function(events){
        if (events && this._eventObj) 
        {
          // Unbinding previous Events
          if (this._parameters.bind){
            var oldEvents = this._parameters.bind;
            for (var a in oldEvents) if (oldEvents.hasOwnProperty(a)) 
              // TODO: Remove jQuery dependency
              jQuery(this._eventObj).unbind(a,oldEvents[a]);
          }
 
        this._parameters.bind = events;
        for (var a in events) if (events.hasOwnProperty(a)) 
          // TODO: Remove jQuery dependency
          jQuery(this._eventObj).bind(a,events[a]);
        }
      },
 
      _Loader:(function()
      {
        if (IE)
          return function() {
            var width=this._img.width;
            var height=this._img.height;
            this._imgWidth = width;
            this._imgHeight = height; 
            this._img.parentNode.removeChild(this._img);
 
            this._vimage = this.createVMLNode('image');
            this._vimage.src=this._img.src;
            this._vimage.style.height=height+"px";
            this._vimage.style.width=width+"px";
            this._vimage.style.position="absolute"; // FIXES IE PROBLEM - its only rendered if its on absolute position!
            this._vimage.style.top = "0px";
            this._vimage.style.left = "0px";
            this._aspectW = this._aspectH = 1;
 
            /* Group minifying a small 1px precision problem when rotating object */
            this._container = this.createVMLNode('group');
            this._container.style.width=width;
            this._container.style.height=height;
            this._container.style.position="absolute";
            this._container.style.top="0px";
            this._container.style.left="0px";
            this._container.setAttribute('coordsize',width-1+','+(height-1)); // This -1, -1 trying to fix ugly problem with small displacement on IE
            this._container.appendChild(this._vimage);
 
            this._rootObj.appendChild(this._container);
            this._rootObj.style.position="relative"; // FIXES IE PROBLEM
            this._rootObj.style.width=width+"px";
            this._rootObj.style.height=height+"px";
            this._rootObj.setAttribute('id',this._img.getAttribute('id'));
            this._rootObj.className=this._img.className;			
            this._eventObj = this._rootObj;	
            var parameters;
            while (parameters = this._onLoadDelegate.shift()) {
              this._handleRotation(parameters, true);	
            }
          }
          else return function () {
            this._rootObj.setAttribute('id',this._img.getAttribute('id'));
            this._rootObj.className=this._img.className;
 
            this._imgWidth=this._img.naturalWidth;
            this._imgHeight=this._img.naturalHeight;
            var _widthMax=Math.sqrt((this._imgHeight)*(this._imgHeight) + (this._imgWidth) * (this._imgWidth));
            this._width = _widthMax * 3;
            this._height = _widthMax * 3;
 
            this._aspectW = this._img.offsetWidth/this._img.naturalWidth;
            this._aspectH = this._img.offsetHeight/this._img.naturalHeight;
 
            this._img.parentNode.removeChild(this._img);	
 
 
            this._canvas=document.createElement('canvas');
            this._canvas.setAttribute('width',this._width);
            this._canvas.style.position="relative";
            this._canvas.style.left = -this._img.height * this._aspectW + "px";
            this._canvas.style.top = -this._img.width * this._aspectH + "px";
            this._canvas.Wilq32 = this._rootObj.Wilq32;
 
            this._rootObj.appendChild(this._canvas);
            this._rootObj.style.width=this._img.width*this._aspectW+"px";
            this._rootObj.style.height=this._img.height*this._aspectH+"px";
            this._eventObj = this._canvas;
 
            this._cnv=this._canvas.getContext('2d');
            var parameters;
            while (parameters = this._onLoadDelegate.shift()) {
              this._handleRotation(parameters, true);	
            }
          }
      })(),
 
      _animateStart:function()
      {	
        if (this._timer) {
          clearTimeout(this._timer);
        }
        this._animateStartTime = +new Date;
        this._animateStartAngle = this._angle;
        this._animate();
      },
      _animate:function()
      {
        var actualTime = +new Date;
        var checkEnd = actualTime - this._animateStartTime > this._parameters.duration;
 
        // TODO: Bug for animatedGif for static rotation ? (to test)
        if (checkEnd && !this._parameters.animatedGif) 
        {
          clearTimeout(this._timer);
        }
        else 
        {
          if (this._canvas||this._vimage||this._img) {
            var angle = this._parameters.easing(0, actualTime - this._animateStartTime, this._animateStartAngle, this._parameters.animateTo - this._animateStartAngle, this._parameters.duration);
            this._rotate((~~(angle*10))/10);
          }
          if (this._parameters.step) {
            this._parameters.step(this._angle);
          }
          var self = this;
          this._timer = setTimeout(function()
          {
            self._animate.call(self);
          }, 10);
        }
 
      // To fix Bug that prevents using recursive function in callback I moved this function to back
      if (this._parameters.callback && checkEnd){
        this._angle = this._parameters.animateTo;
        this._rotate(this._angle);
        this._parameters.callback.call(this._rootObj);
      }
      },
 
      _rotate : (function()
      {
        var rad = Math.PI/180;
        if (IE)
          return function(angle)
        {
          this._angle = angle;
          this._container.style.rotation=(angle%360)+"deg";
          this._vimage.style.top = -(this._rotationCenterY - this._imgHeight/2) + "px";
          this._vimage.style.left = -(this._rotationCenterX - this._imgWidth/2) + "px";
          this._container.style.top = this._rotationCenterY - this._imgHeight/2 + "px";
          this._container.style.left = this._rotationCenterX - this._imgWidth/2 + "px";
 
        }
          else if (supportedCSS)
          return function(angle){
            this._angle = angle;
            this._img.style[supportedCSS]="rotate("+(angle%360)+"deg)";
            this._img.style[supportedCSSOrigin]=this._parameters.center.join(" ");
          }
          else 
            return function(angle)
          {
            this._angle = angle;
            angle=(angle%360)* rad;
            // clear canvas	
            this._canvas.width = this._width;//+this._widthAdd;
            this._canvas.height = this._height;//+this._heightAdd;
 
            // REMEMBER: all drawings are read from backwards.. so first function is translate, then rotate, then translate, translate..
            this._cnv.translate(this._imgWidth*this._aspectW,this._imgHeight*this._aspectH);	// at least center image on screen
            this._cnv.translate(this._rotationCenterX,this._rotationCenterY);			// we move image back to its orginal 
            this._cnv.rotate(angle);										// rotate image
            this._cnv.translate(-this._rotationCenterX,-this._rotationCenterY);		// move image to its center, so we can rotate around its center
            this._cnv.scale(this._aspectW,this._aspectH); // SCALE - if needed ;)
            this._cnv.drawImage(this._img, 0, 0);							// First - we draw image
          }
 
      })()
      }
 
      if (IE)
      {
        Wilq32.PhotoEffect.prototype.createVMLNode=(function(){
          document.createStyleSheet().addRule(".rvml", "behavior:url(#default#VML)");
          try {
            !document.namespaces.rvml && document.namespaces.add("rvml", "urn:schemas-microsoft-com:vml");
            return function (tagName) {
              return document.createElement('<rvml:' + tagName + ' class="rvml">');
            };
          } catch (e) {
            return function (tagName) {
              return document.createElement('<' + tagName + ' xmlns="urn:schemas-microsoft.com:vml" class="rvml">');
            };
          }		
        })();
      }
 
})(jQuery);